### ISSUE TITLE
Support %% cell magic syntax and more specifically %%HTML

### ISSUE THREAD

**Issue by lbustelo** [IRkernel/IRkernel#260]
In order to implement support for https://github.com/jupyter-incubator/declarativewidgets, we need support for the %%HTML magic. In declarativewidgets user create their interactive areas using HTML markup.

**Comment by rgbkrk** [comment]
Will this conflict with operators like `%*%` and `%+%` (as used with ggplot)?

**Comment by rgbkrk** [comment]
Ah, and `%%` is modulus in R.

**Comment by rgbkrk** [comment]
Seems like that can be parsed off first, not sure what folks think about it.

**Comment by JanSchulz** [comment]
Wouldn't it be easier to ask the declarative widget people to make the way to submit HTML configurable?

**Comment by takluyver** [comment]
I'm the person who has implemented the input transformation machinery which allows magics in IPython, and I'm categorically against trying to copy it for R. Python is a good language for that kind of thing, and it's still tricky. R would be a *terrible* language for that kind of machinery.

**Comment by rgbkrk** [comment]
My main disagreement with magics is that they're not portable when exporting to vanilla scripts.

That being said, people find them useful.

**Comment by lbustelo** [comment]
I wish that some magics were fully evaluated by the client... but that is not how it works.

@JanSchulz declarativewidgets is not really imposing usage of %%HTML, it is just an easy way to author cells that contain HTML (lots of HTML). We could just say that for R you have to bundle the HTML using `IRdisplay::display_html` but I think that is not the ideal experience for users.

**Comment by rgbkrk** [comment]
@lbustelo Definitely hits on that point you made earlier about a cell type being HTML, instead of having that as a magic.

**Comment by takluyver** [comment]
Does it work if you just put the HTML in a markdown cell? HTML shouldn't need to go through the kernel.

**Comment by lbustelo** [comment]
@takluyver Would be awesome if it did, and probably a great place for this content, but it seems that it has a limited white lists of tags.

**Comment by flying-sheep** [comment]
maybe that should be changed. what a man and his notebook do behind closed doors should be nobody’s business :wink:

**Comment by takluyver** [comment]
@lbustelo do you want to open on the notebook repo to work something out for this? Whether it's expanding what's possible in markdown cells or creating a new option, I think there should be a way to write arbitrary HTML without involving the kernel. It can't be a security concern, because you can already have arbitrary HTML in outputs.

**Comment by rgbkrk** [comment]
> It can't be a security concern, because you can already have arbitrary HTML in outputs.

Yeah, and it would be part of the sanitization and security mechanisms. A+

This would be a new cell type for nbformat to talk about on Jupyter. :+1:

**Comment by lbustelo** [comment]
@rgbkrk I already got some pushback about new cell types once before.

@takluyver It would be an interesting investigation to not only allow the tags to fall on the output area of a markdown cell, but also if they can communicate with the kernel once they land on the page. I'll do some digging around and open something on notebook repo once I'm more educated. I agree that the roundtrip for both HTML and Javascript is unnecessary.

**Comment by takluyver** [comment]
I don't think there's anything preventing HTML in a markdown cell from communicating with the kernel.

**Comment by rgbkrk** [comment]
> I already got some pushback about new cell types once before.

At the very least, I think you have a really well defined use case and now with N many languages having to reason about `%%html` like dynamics, it's worth bringing up again. I'll bring it up.

/cc @jdfreder

**Comment by flying-sheep** [comment]
result of digging: Jupyter notebook uses [marked](https://github.com/chjj/marked).

* configured [here](https://github.com/jupyter/notebook/blob/8d4a8ee0c7c2f83560d45e4386b7e04c919e4bad/notebook/static/notebook/js/notebook.js#L83)
* used [here](https://github.com/jupyter/notebook/blob/7e6b50815e2626df8a261dfb2f42f5d12a37fb5c/notebook/static/notebook/js/textcell.js#L274)

marked has a [sanitize](https://github.com/chjj/marked#sanitize) option, per default `false`, which stips *all* HTML

therefore (i guess) jupyter notebook uses [`security.sanitize_html`](https://github.com/jupyter/notebook/blob/932f85c7593b2eada7a2095bd1f30d3f6ea48115/notebook/static/base/js/security.js#L73) ([here](https://github.com/jupyter/notebook/blob/7e6b50815e2626df8a261dfb2f42f5d12a37fb5c/notebook/static/notebook/js/textcell.js#L276). let’s make it not do so.

**Comment by ellisonbg** [comment]
I know probably won't satisfy all the usage cases, but non-python kernels can always push HTML without using magics with regular display machinery. For algorithmic generation of HTML, that works fine. Where it breaks down is when users want to type HTML themselves.

**Comment by ellisonbg** [comment]
I agree with @takluyver that doing the input transformations in R might be a bit painful - even in Python that are pretty nasty...

**Comment by ellisonbg** [comment]
The other thing that is lurking here is that the pain of magics in some languages isn't just about `%%html`  but rather *all* cell and line magics. Addressing the pain of `%%html` but ignoring the larger problem of magics still leaves a lot of pain for other kernels to deal with...don't have ideas on how to solve that...

**Comment by JanSchulz** [comment]
Inline magics are hard, but cell level magics would be nice: Notebook issue about generic cell magics for the notebook: https://github.com/jupyter/notebook/issues/746

**Comment by fperez** [comment]
Very early on, before we had a notebook but when we were working on the Qt console, I remember we agonized about adding yet another syntax for client-side magics.  I think I proposed `:foo` as the syntax (i.e., `:` would be the 'special syntax that stays client-side' escape).

In the end we decided against it, worried that `!` and `%` was already enough special parsing of the input to add yet more cases...

I'll follow up further on jupyter/notebook#1123...

**Comment by janschulz** [comment]
IMO this should be closed as wontfix:

We currently have no plans to add an R specific magic system. If users want to display something, the IRdisplay functions should be used.

Maybe this should become a FAQ in the docs and on the website...

**Comment by takluyver** [comment]
Yep, I think we agreed that we need a way to get HTML/JS on the page without going through the kernel, whether that's a new cell type, lifting the restrictions on markdown cells, or something else.

**Comment by flying-sheep** [comment]
added IRkernel/irkernel.github.io#26

**Comment by takluyver** [comment]
Thanks @flying-sheep

**Comment by abalter** [comment]
@rgbkrk I think you really hit the nail on the head with HTML cell type. That was my goal for this little dabbling. Gosh I hope to get back to it.

https://github.com/abalter/polyglottus

@lbustelo -- what is the reason for NOT wanting more cell types? 

What I think the world needs is
1. Cell kernels of any time
1. Separate (named) kernels for specific cells (if wanted)
1. Unified syntax for passing variables between kernels (cells)
